### ISSUE TITLE
gzip chokes due to very old atime

### ISSUE THREAD

**Issue by kernelt** [GoogleCloudPlatform/gcsfuse#148]
Hi Aaron. We noticed an error if we are using a gzip command in a bucket-folder.
```
touch test-gzip.file
gzip test-gzip.file
gzip: test-gzip.file.gz: Invalid argument
```
The test-gzip.file.gz file was created, but we see an error and the file can't be unpacked - the archive is broken. Is it possible to fix it? See detail logs of a gcsfuse below.
```
touch test-gzip.file
fuse_debug: Op 0x0000000a        connection.go:395] <- GetInodeAttributes (inode 1)
fuse_debug: Op 0x0000000a        connection.go:474] -> OK
fuse_debug: Op 0x0000000b        connection.go:395] <- LookUpInode (parent 1, name "test-gzip.file")
fuse_debug: Op 0x0000000b        connection.go:476] -> Error: "no such file or directory"
fuse_debug: Op 0x0000000c        connection.go:395] <- CreateFile
fuse_debug: Op 0x0000000c        connection.go:474] -> OK
fuse_debug: Op 0x0000000d        connection.go:395] <- FlushFile (inode 3)
fuse_debug: Op 0x0000000d        connection.go:474] -> OK
fuse_debug: Op 0x0000000e        connection.go:395] <- SetInodeAttributes (inode 3, atime 2016-01-09 15:28:50.192340982 +0000 UTC, mtime 2016-01-09 15:28:50.192340982 +0000 UTC)
fuse_debug: Op 0x0000000e        connection.go:474] -> OK
fuse_debug: Op 0x0000000f        connection.go:395] <- unknown (inode 3, opcode 24)
fuse_debug: Op 0x0000000f        connection.go:476] -> Error: "function not implemented"
fuse_debug: Op 0x00000010        connection.go:395] <- FlushFile (inode 3)
fuse_debug: Op 0x00000010        connection.go:474] -> OK
fuse_debug: Op 0x00000011        connection.go:395] <- ReleaseFileHandle
fuse_debug: Op 0x00000011        connection.go:474] -> OK

gzip test-gzip.file
fuse_debug: Op 0x00000012        connection.go:395] <- GetInodeAttributes (inode 1)
fuse_debug: Op 0x00000012        connection.go:474] -> OK
fuse_debug: Op 0x00000013        connection.go:395] <- LookUpInode (parent 1, name "test-gzip.file")
fuse_debug: Op 0x00000013        connection.go:474] -> OK
fuse_debug: Op 0x00000014        connection.go:395] <- OpenFile (inode 3)
fuse_debug: Op 0x00000014        connection.go:474] -> OK
fuse_debug: Op 0x00000015        connection.go:395] <- LookUpInode (parent 1, name "test-gzip.file.gz")
fuse_debug: Op 0x00000015        connection.go:476] -> Error: "no such file or directory"
fuse_debug: Op 0x00000016        connection.go:395] <- CreateFile
fuse_debug: Op 0x00000016        connection.go:474] -> OK
fuse_debug: Op 0x00000017        connection.go:395] <- ReadFile (inode 3, handle 1, offset 0, 4096 bytes)
fuse_debug: Op 0x00000017        connection.go:474] -> OK
fuse_debug: Op 0x00000018        connection.go:395] <- unknown (inode 4, opcode 22)
fuse_debug: Op 0x00000018        connection.go:476] -> Error: "function not implemented"
fuse_debug: Op 0x00000019        connection.go:395] <- WriteFile (inode 4, handle 2, offset 0, 35 bytes)
fuse_debug: Op 0x00000019        connection.go:474] -> OK
fuse_debug: Op 0x0000001a        connection.go:395] <- FlushFile (inode 3)
fuse_debug: Op 0x0000001a        connection.go:474] -> OK
fuse_debug: Op 0x0000001b        connection.go:395] <- ReleaseFileHandle
fuse_debug: Op 0x0000001c        connection.go:395] <- SetInodeAttributes (inode 4)
fuse_debug: Op 0x0000001b        connection.go:474] -> OK
fuse_debug: Op 0x0000001c        connection.go:474] -> OK
fuse_debug: Op 0x0000001d        connection.go:395] <- SetInodeAttributes (inode 4, mode -rwxrwxr-x)
fuse_debug: Op 0x0000001d        connection.go:474] -> OK
fuse_debug: Op 0x0000001e        connection.go:395] <- FlushFile (inode 4)
fuse_debug: Op 0x0000001e        connection.go:474] -> OK
fuse_debug: Op 0x0000001f        connection.go:395] <- ReleaseFileHandle
fuse_debug: Op 0x00000020        connection.go:395] <- GetInodeAttributes (inode 1)
fuse_debug: Op 0x00000020        connection.go:474] -> OK
fuse_debug: Op 0x0000001f        connection.go:474] -> OK
fuse_debug: Op 0x00000021        connection.go:395] <- LookUpInode (parent 1, name "test-gzip.file")
fuse_debug: Op 0x00000021        connection.go:474] -> OK
fuse_debug: Op 0x00000022        connection.go:395] <- Unlink
fuse_debug: Op 0x00000022        connection.go:474] -> OK
fuse_debug: Op 0x00000023        connection.go:395] <- ForgetInode (inode 3)
fuse_debug: Op 0x00000023        connection.go:474] -> OK
```

**Comment by jacobsa** [comment]
I can reproduce this on Ubuntu Wily (but not OS X). It doesn't appear to relate
to any crazy GCS content sniffing, because it happens even with the fake
bucket. Looking.

**Comment by jacobsa** [comment]
`strace` shows no errors received by gzip from the kernel. This is something internal to gzip. Sigh, compiling gzip with printf debugging...

**Comment by jacobsa** [comment]
printf debugging shows that this happens in the warning logging after the call to `fdutimens` in the gzip 1.6 source code, as downloaded [here](http://gnu.mirror.iweb.com/gzip/gzip-1.6.tar.gz). (Ugh, I can't find somewhere that hosts the versioned source in a browsable form.)

**Comment by jacobsa** [comment]
For the record, here's what `gzip` does after loading libraries and masking signals:

```
open("/home/jacobsa/mp/foo", O_RDONLY|O_NOCTTY|O_NONBLOCK|O_NOFOLLOW) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=0, ...}) = 0
rt_sigprocmask(SIG_BLOCK, [HUP INT PIPE TERM XCPU XFSZ], [], 8) = 0
open("/home/jacobsa/mp/foo.gz", O_WRONLY|O_CREAT|O_EXCL, 0600) = 4
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
read(3, "", 65536)                      = 0
write(4, "\37\213\10\10\200\334\222V\0\3foo\0\3\0\0\0\0\0\0\0\0\0", 24) = 24
close(3)                                = 0
write(2, "gzip: ", 6gzip: )                   = 6
write(2, "/home/jacobsa/mp/foo.gz: Invalid"..., 42/home/jacobsa/mp/foo.gz: Invalid argument
) = 42
fchown(4, 1000, 1000)                   = 0
fchmod(4, 0644)                         = 0
close(4)                                = 0
rt_sigprocmask(SIG_BLOCK, [HUP INT PIPE TERM XCPU XFSZ], [], 8) = 0
unlink("/home/jacobsa/mp/foo")          = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
lseek(0, 0, SEEK_CUR)                   = -1 ESPIPE (Illegal seek)
close(0)                                = 0
close(1)                                = 0
close(2)                                = 0
exit_group(2)                           = ?
+++ exited with 2 +++
```

**Comment by jacobsa** [comment]
Ah, gzip is failing in `fdutimens` (in `lib/utimens.c`) due to `validate_timespec` failing due to the first `timespec` being seen as invalid. Running `strace -v` shows why, in the result from `fstat`:

    st_atime=1754/08/30-22:43:42, st_mtime=2016/01/10-22:42:37, st_ctime=2016/01/10-22:42:37

`gzip` doesn't like that value for `st_atime`.

**Comment by jacobsa** [comment]
Indeed, if I return an atime of `time.Now()`, `gzip` doesn't complain.

**Comment by jacobsa** [comment]
Solution:

1. Specify in the semantics that atime will be set to something reasonable.
2. Add tests for this.
3. In reality, set it to be equal to mtime.

**Comment by jacobsa** [comment]
This should be fixed in version 0.15.1. Please let me know.

**Comment by kernelt** [comment]
now, it works! thanks!
